### PR TITLE
[MCLEAN-111] Fix fast delete feature for filesets

### PR DESCRIPTION
**Pull request by wilx**
Pass `GlobSelector` instance into `cleaner.delete()` only if it matters. Passing `GlobSelector` always breaks the fast delete feature.

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MCLEAN-111) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Format the pull request title like `[MCLEAN-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MCLEAN-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [X] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by wilx** [comment]
JIRA: https://issues.apache.org/jira/projects/MCLEAN/issues/MCLEAN-102

**Comment by slawekjaranowski** [comment]
@wilx thanks
Is it possible to make an IT to be sure for regressions with next changes?

**Comment by wilx** [comment]
@slawekjaranowski I have added the test.

File: src/it/fast-delete/invoker.properties [hunk 1]
```diff
@@ -0,0 +1 @@
+invoker.goals = install clean -Dorg.slf4j.simpleLogger.showThreadName=true -X -e
```
Comment (slawekjaranowski) [inline src/it/fast-delete/invoker.properties:1]: missed license header 😄

Comment (wilx) [reply src/it/fast-delete/invoker.properties:18]: Fixed.

**Comment by chabala** [comment]
Is this a symptom of `GlobSelector` not working on Windows at all?

**Comment by slawekjaranowski** [comment]
> Is this a symptom of `GlobSelector` not working on Windows at all?

Good point - it should be also checked.

**Comment by wilx** [comment]
> Is this a symptom of `GlobSelector` not working on Windows at all?

The issue is not with `GlobSelector`. The issue is that the `GlobSelector` instance passed into `cleaner.delete()` stops the fast method from taking effect: https://github.com/apache/maven-clean-plugin/blob/06876c185555de04040448d7fb7e5287b96307de/src/main/java/org/apache/maven/plugins/clean/Cleaner.java#L122-L127

**Comment by slawekjaranowski** [comment]
So it looks like when we use includes/excludes fastDelete will not works - correct?

File: src/it/fast-delete/pom.xml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	<properties>
+		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+	</properties>
+
+	<build>
+		<plugins>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-clean-plugin</artifactId>
+				<version>@pom.version@</version>
```
Comment (slawekjaranowski) [inline src/it/fast-delete/pom.xml:43]: should be `@project.version@` ... in other tests we need fix also - but in separate PR 😄

**Comment by wilx** [comment]
> So it looks like when we use includes/excludes fastDelete will not works - correct?

Yes and it makes sense. The fast delete moves the entire directory. If you want to delete only some files, you cannot move the directory because the files that should stay there would not.

File: src/it/fast-delete/pom.xml (hunk 2 above)
Comment (wilx) [reply src/it/fast-delete/pom.xml]: Fixed.

**Comment by slawekjaranowski** [comment]
> > So it looks like when we use includes/excludes fastDelete will not works - correct?
> 
> Yes and it makes sense. The fast delete moves the entire directory. If you want to delete only some files, you cannot move the directory because the files that should stay there would not.

Yes make sense.

Can you update documentation?
https://maven.apache.org/plugins/maven-clean-plugin/examples/delete_additional_files.html
(source in project src/site/apt/examples/delete_additional_files.apt.vm)

We should explain such behaviors.

**Comment by chabala** [comment]
> > So it looks like when we use includes/excludes fastDelete will not works - correct?
> 
> Yes and it makes sense. The fast delete moves the entire directory. If you want to delete only some files, you cannot move the directory because the files that should stay there would not.

I agree. Looks like fast delete has always required `GlobSelector` to be null, since it was added in #6. So, it has never worked for this `Fileset` parsing path, only the directory parsing path above it at: https://github.com/apache/maven-clean-plugin/blob/f9da326e1b1ad4648d24e646b51cc7bc0a1baad7/src/main/java/org/apache/maven/plugins/clean/CleanMojo.java#L250-L254

**Comment by wilx** [comment]
> > > So it looks like when we use includes/excludes fastDelete will not works - correct?
> > 
> > 
> > Yes and it makes sense. The fast delete moves the entire directory. If you want to delete only some files, you cannot move the directory because the files that should stay there would not.
> 
> Yes make sense.
> 
> Can you update documentation? https://maven.apache.org/plugins/maven-clean-plugin/examples/delete_additional_files.html (source in project src/site/apt/examples/delete_additional_files.apt.vm)
> 
> We should explain such behaviors.

I have added a note to the `delete_additional_files.apt.vm`.

**Comment by slawekjaranowski** [comment]
@wilx  - thanks for fix

**Comment by jira-importer** [comment]
Resolve #176

**Pull request merged**